realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java

```diff
@@ -482,8 +499,13 @@ private void executePendingRowQueries() {
     @SuppressWarnings("unused")
     private static void runMigrationCallback(long nativeSharedRealmPtr, OsRealmConfig osRealmConfig, MigrationCallback callback,
                                              long oldVersion) {
-        callback.onMigrationNeeded(new SharedRealm(nativeSharedRealmPtr, osRealmConfig), oldVersion,
-                osRealmConfig.getRealmConfiguration().getSchemaVersion());
+        SharedRealm sharedRealm = new SharedRealm(nativeSharedRealmPtr, osRealmConfig);
```

Comment:
Not sure I understand the reason for making this change?

Reply:
complex story.
take the migration for example, ObjectStore will pass a `SharedRealm` instance to the callback
to make the life control easier, i created a new pointer from that instance and pass it to java, so it can be managed like a normal `SharedRealm`, user will never meet a invalid pointer of `SharedRealm` even when he passed that `SharedRealm` instance out.

But then it is a problem when exception happens in migration block. after exception thrown, `getInstance()` hasn't return anything yet, so user doesn't have to call `close()`. But there is a `SharedRealm` pointer created in the migration callback which is waiting for the GC to be deleted. That means, from OS's RealmCoordinator pointer of view, there is still one `SharedRealm` is not dealloced. -- call `deleteRealm()` at this time will throw.

But i think i still miss one thing in this modification, `SharedRealm.isClosed()` should return true after `delete()` called to avoid use pass the `Realm`/`DyanmicRealm` instance out of the callback and see a deleted pointer.